src/Net/HttpServer.cs

```diff
@@ -633,22 +633,23 @@ private void HandleDisconnect()
 
         private void OnClearCache()
         {
-            var keys = _processedPokemon.Keys.ToList();
-            for (var i = 0; i < keys.Count; i++)
+            List<string> expiredEncounters;
+            
+            lock (_processedPokemon)
             {
-                var encounterId = keys[i];
-                if (encounterId != null)
-                {
-                    var scannedPokemon = _processedPokemon[encounterId];
+                expiredEncounters = _processedPokemon.Where(pair => pair.Value.IsExpired).Select(pair => pair.Key).ToList();
```

Comment:
```suggestion
                expiredEncounters = _processedPokemon.Where(pair => pair.Value.IsExpired).Select(pair => pair.Key);
```
Since `Select` returns an enumerable and we aren't using any list specific methods we could probably omit `ToList`?

Reply:
The problem is that the enumerable returned from a `Dictionary` represents a _view_ of that Dictionary that changes whenever the Dictionary does, meaning once it leaves the `lock` block, that enumerable could change if another thread modifies the Dictionary and the results would be invalid. You're also technically not supposed to enumerate an `IEnumerable` more than once. Using `ToList` materializes all the expired keys to an actual list there so that once you leave the `lock` block, that collection won't be modified by another thread while logging the expired encounters

Reply:
Gotcha, thanks for the explanation.